packages/core/content-manager/admin/src/pages/EditView/components/FormInputs/Relations/RelationModal.tsx

```diff
@@ -57,91 +61,292 @@ const CustomModalContent = styled(Modal.Content)`
+      payload: { hasUnsavedChanges: boolean };
+    };
+
+function reducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'GO_TO_RELATION':
+      if (state.hasUnsavedChanges && !action.payload.shouldBypassConfirmation) {
+        return { ...state, confirmDialogIntent: 'open' };
+      }
+
+      return {
+        ...state,
+        documentHistory: [...state.documentHistory, action.payload.document],
+        confirmDialogIntent: null,
+        isModalOpen: true,
+      };
+    case 'GO_BACK':
+      if (state.hasUnsavedChanges && !action.payload.shouldBypassConfirmation) {
+        return { ...state, confirmDialogIntent: 'back' };
+      }
+
+      return {
+        ...state,
+        documentHistory: state.documentHistory.slice(0, state.documentHistory.length - 1),
+        confirmDialogIntent: null,
+      };
+    case 'GO_FULL_PAGE':
+      if (state.hasUnsavedChanges) {
+        return { ...state, confirmDialogIntent: 'navigate' };
+      }
+
+      return {
+        ...state,
+        confirmDialogIntent: null,
+      };
+    case 'CANCEL_CONFIRM_DIALOG':
+      return {
+        ...state,
+        confirmDialogIntent: null,
+      };
+    case 'CLOSE_MODAL':
+      if (state.hasUnsavedChanges && !action.payload.shouldBypassConfirmation) {
+        return { ...state, confirmDialogIntent: 'close' };
+      }
+
+      return {
+        ...state,
+        documentHistory: [],
+        confirmDialogIntent: null,
+        isModalOpen: false,
+      };
+    case 'SET_HAS_UNSAVED_CHANGES':
+      return {
+        ...state,
+        hasUnsavedChanges: action.payload.hasUnsavedChanges,
+      };
+    default:
+      return state;
   }
-);
+}
 
-const RelationModalForm = ({ relation, triggerButtonLabel }: RelationModalProps) => {
-  const navigate = useNavigate();
-  const { pathname, search } = useLocation();
-  const { formatMessage } = useIntl();
+interface RelationModalContextValue {
+  state: State;
+  dispatch: React.Dispatch<Action>;
+  rootDocumentMeta: DocumentMeta;
+  currentDocumentMeta: DocumentMeta;
+  currentDocument: ReturnType<UseDocument>;
+  onPreview?: () => void;
+}
 
-  const [triggerRefetchDocument] = useLazyGetDocumentQuery();
+const [RelationModalProvider, useRelationModal] =
+  createContext<RelationModalContextValue>('RelationModal');
 
-  const currentDocument = useDocumentContext('RelationModalForm', (state) => state.document);
-  const rootDocumentMeta = useDocumentContext(
-    'RelationModalForm',
-    (state) => state.rootDocumentMeta
-  );
-  const currentDocumentMeta = useDocumentContext('RelationModalForm', (state) => state.meta);
-  const changeDocument = useDocumentContext('RelationModalForm', (state) => state.changeDocument);
-  const documentHistory = useDocumentContext('RelationModalForm', (state) => state.documentHistory);
-  const setDocumentHistory = useDocumentContext(
-    'RelationModalForm',
-    (state) => state.setDocumentHistory
-  );
+const FormWatcher = () => {
```

Comment:
Need to check if I can move this inside another existing component that has access to the form context. Because I don't love having a component that doesn't return any UI